### DISCUSSION TITLE
Why doesn't Bootstrap 5 make better use of CSS color variables?

### DISCUSSION THREAD

**Discussion by MontyHu** [twbs/bootstrap#32388] [General] [answered]
I notice that Bootstrap 5 makes more use of CSS variables in the compiled .css file, which is great. But for some reason the color variables aren't being used. For example, at the top of the bootstrap.css file is: 

```
:root {
  --bs-primary: #007bff;
}
```

That primary color is used in lots of places further down in the .css file, such as the link color. But instead of using `color: var(--bs-primary);` the color is hard-coded into the definition: 

```
a {
  color: #007bff;
  text-decoration: underline;
}
```

Wouldn't it make more sense to make those colors CSS variables? It would make customizing the default bootstrap.css file a lot easier for templates when you want to allow someone to change the accent/link color without having to create multiple versions of the bootstrap.css file using sass, node, npm and gulp when you only want to change a few colors. 

Currently what I do is create an override.css file that redefines every css definition that uses `#007bff` to `var(--bs-primary)` like this: 

```
a {
  color: var(--bs-primary);
}
.btn-primary {
  background-color: var(--bs-primary);
}
.btn-primary.disabled, .btn-primary:disabled {
  background-color: var(--bs-primary);
  border-color: var(--bs-primary); 
}
etc... 
```

But that is a bit tedious, and when updates are made to Bootstrap, I have to trawl through the bootstrap.css file again to see if any new definitions were added that use any of the colors defined in :root so I can create overrides using the variable. 

Is there a reason why css variables for colors aren't used in the compiled .css file?

**Comment by mdo** [comment, marked as answer]
When we started v5's development, IE11 was still on the table, and it doesn't support CSS variables. We dropped support for IE11 a few months ago and have started incorporating CSS vars across the board. They're not a replacement for Sass variables though, so it'll take time for us to implement.

Beyond that, we do a _ton_ of variable reassignment (e.g., `$btn-primary-bg: $primary`) and doing that in CSS variables in addition to Sass variables feels like the wrong approach. Might take us a little longer to work it all out and our approach, but we'll iterate on our way to get there.

I'd suggest embracing Sass and using Bootstrap that way if you find yourself needing to customize things extensively.

**Reply by MontyHu** [reply]
Thanks, Mark. Glad to hear that CSS variables are being incorporated into Bootstrap more. I have been using Sass to make customizations, although I find all the Sass/NPM/Gulp stuff very confusing. I still can't figure out how to incorporate Autoprefixer into the process so that certain form elements don't break in BS5 after compiling, as I explained in [this thread](https://github.com/twbs/bootstrap/discussions/32369). One day I'll figure it out.

**Reply by mdo** [reply]
Have you checked out the npm starter project I created not too long ago? https://github.com/twbs/bootstrap-npm-starter

It does as much as possible for Sass/CSS and light customization, without the need for Gulp. My whole thing there is front-end dev tooling is so complicated, so I try to keep things focused on as few tools as possible.

**Reply by MontyHu** [reply]
I just checked it out. I don't know how to use Github templates, but I guess it's another thing I'll have to learn, especially if it removes some of the complexity of customizing Bootstrap. Do you think this would eventually be available to work with Bootstrap 5?

**Reply by mdo** [reply]
You can also just download the repo's zip file. The template repo just allows you to copy it to your own GitHub username or organization so you can modify it however you want.

**Comment by mdo** [comment]
Some stuff coming in #32424.

**Comment by dbackeus** [comment]
Would just like to second that Bootstrap adopting standard CSS as far as possible would be very nice indeed. After reading https://getbootstrap.com/docs/5.0/customize/css-variables/ I was quite surprised to realize these variables were not actually useful for customizing bootstrap's default theme.

Worth keepingin mind that the [buildless](https://buildless.site/) movement is slowly taking shape and a growing number of developers are now looking forward to a future free of build-tool complexity.

**Reply by MontyHu** [reply]
Yes, a buildless Bootstrap that can be customized more easily with CSS variables is still very much needed! I was hoping that would become more possible with Boostrap 5, but, I still have to create an override.css file that replaces hard-coded colors with CSS variables to make it easier to customize colors in BS5. 

It would be so great to not have to use npm, gulp and sass just to customize a few color variables in Bootstrap. But I think things are getting there, I'm hoping with BS6 we'll be able to use CSS variables to customize all colors and some other basic things. This would make BS a lot better for creating templates that offer different color schemes, for example.
